src/Common/src/TypeSystem/Interop/IL/Marshaller.cs

```diff
@@ -1878,6 +1880,28 @@ protected override void EmitCleanupManaged(ILCodeStream codeStream)
         }
     }
 
+    class HandleRefMarshaller : Marshaller
+    {
+        protected override void AllocAndTransformManagedToNative(ILCodeStream codeStream)
+        {
+            LoadManagedAddr(codeStream);
+            codeStream.Emit(ILOpcode.ldfld, _ilCodeStreams.Emitter.NewToken(InteropTypes.GetHandleRef(Context).GetKnownField("_handle")));
+            StoreNativeValue(codeStream);
+        }
+
+        protected override void TransformNativeToManaged(ILCodeStream codeStream)
```

Comment:
May want to override `protected override void TransformManagedToNative(ILCodeStream codeStream)` as well and throw InvalidProgramException() from it. I suspect that it may get called in some cases.